singletons-th/src/Data/Singletons/TH/Promote/Monad.hs

```diff
@@ -115,3 +123,102 @@ promoteMDecs :: OptionsMonad q => [Dec] -> PrM [DDec] -> q [DDec]
 promoteMDecs locals thing = do
   (decs1, decs2) <- promoteM locals thing
   return $ decs1 ++ decs2
+
+{-
+Note [Tracking local variables]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Handling local variables in singletons-th requires some care. There are two
+sorts of local variables that singletons-th tracks:
+
+1. Lambda-bound variables, e.g.,
+
+     f = \x -> x
+     g x = x
+
+   In both `f` and `g`, the variable `x` is considered lambda-bound.
+
+2. Let-bound variables, e.g.,
+
+     h =
+       let x = 42 in
+       x + x
+
+     i = x + x
+       where
+         x = 42
+
+   In both `h` and `i`, the variable `x` is considered let-bound.
+
+Why does singletons-th need to track local variables? It's because they must
+be promoted differently depending on whether they are local or not. Consider:
+
+  j = ... x ...
+
+When promoting the `j` function to a type family `J`, there are two possible
+ways of promoting `x`:
+
+* If `x` is a lambda-bound variable, then `x` must be promoted to the same
+  name, since `x`'s binding site will also use that name. For instance, we
+  would promote this:
+
+    j x = x
+
+  To this:
+
+    type family J x where
+      J x = x
```

Comment:
This actually isn't quite right. Notably, if a lambda-bound variable uses a symbolic name, e.g.,

```hs
j (%%) x y = x %% y
```

Then `j` will be promoted to this:

```hs
type family J ty x y where
  J ty x y = x `ty` y
```

This is because type variables aren't allowed to have symbolic names like `(%%)`. I will update the Note to reflect this.